Add imageview features to JSXGraph output of plots when the jsx_navigation option is false.

drgrice1 opened the pull request:
There are also some changes for consistency between JSXGraqh and TikZ output.  The axes arrows now only point in the positive direction, and extend to the edge of the board. With the the axes label positioning needed to be changed.  Furthermore, the JSXGraph preferred method for setting the `position` option of a label is switched to (see https://jsxgraph.org/docs/symbols/Label.html#position).
    
Also remove the `$name` suffixes to variables in the javascript.  This javascript is scoped and that suffix is unnecessary. The variable names to not need to be distinct between different JSXGraph images with the scoping.  Furthmore, it increases the size of the generated javascript which is not desirable.

Remove the injection of `\t` in the JSXGraph and TikZ output.  Do not attempt to make the output human readable.  That is not a desired goal.  In fact for the JavasScript output the goal should be to make the output as unreadable as possible.  That code is visible to the student, and could be used in some cases to assist in finding the answer.  For the TikZ code, the objective should be to make the result compile with as few characters as possible.  TeX has limits on the number of characters it can handle in a given line and we don't want to press that limit. The important thing is TeX is that newlines (`\n`) are where they need to be.
    
Remove the margin on the JSXGraph div.  That is inconsistent with the other display output modes.

drgrice1 committed:
Minor things that need to be done.

Remove the injection of `\t` in the JSXGraph and TikZ output.  Do not
attempt to make the output human readable.  That is not a desired goal.
In fact for the JavasScript output the goal should be to make the output
as unreadable as possible.  That code is visible to the student, and
could be used in some cases to assist in finding the answer.  For the
TikZ code, the objective should be to make the result compile with as
few characters as possible.  TeX has limits on the number of characters
it can handle in a given line and we don't want to press that limit. The
important thing is TeX is that newlines (`\n`) are where they need to
be.

Remove the margin on the JSXGraph div.  That is inconsistent with the
other display output modes.

drgrice1 committed:
Add imageview features to JSXGraph output of plots when the jsx_navigation option is false.

There are also some changes for consistency between JSXGraqh and TikZ
output.  The axes arrows now only point in the positive direction, and
extend to the edge of the board. With the the axes label positioning
needed to be changed.  Furthermore, the JSXGraph preferred method for
setting the `position` option of a label is switched to (see
https://jsxgraph.org/docs/symbols/Label.html#position).

Also remove the `$name` suffixes to variables in the javascript.  This
javascript is scoped and that suffix is unnecessary. The variable names
to not need to be distinct between different JSXGraph images with the
scoping.  Furthmore, it increases the size of the generated javascript
which is not desirable.

somiaj commented:
Thanks, I am noticing an issue with the imageview magnifier, it seems to change the range on the board. Happens in many examples I tested, but here is one I am currently working with.

```
DOCUMENT();
loadMacros('PGstandard.pl', 'PGML.pl', 'plots.pl',);

$plot = Plot(
    xmin         => -1,
    xmax         => 6,
    ymin         => -1,
    ymax         => 7,
    xtick_delta  => 1,
    ytick_delta  => 1,
    xminor       => 0,
    yminor       => 0,
    xlabel       => '\(x \text{ (ft)}\)',
    ylabel       => '\(y \text{ (ft)}\)',
    xjsx_options => { label => { offset => [ -27, 5 ] } },
);
$plot->add_color('lightgreen', 156, 215, 151);
$plot->add_color('darkgreen',  0,   86,  34);
$plot->add_dataset(
    [ 0, 1 ], [ 3, 1 ], [ 3, 5 ], [ 0, 5 ], [ 0, 1 ],
    color      => 'darkgreen',
    fill       => 'self',
    fill_color => 'lightgreen',
);

for ('Tikz', 'JSXGraph') {
    $plot->image_type($_);
    $image{$_} = image($plot);
}

BEGIN_PGML
**Tikz Image:**

[$image{Tikz}]*

**JSXGraph Image:**

[$image{JSXGraph}]*

END_PGML
ENDDOCUMENT();
```

Basically the board extends beyond the specified range, and also ends up drawing the rectangle slightly off its desired position.

![myscrot](https://github.com/user-attachments/assets/76ad5ff9-545c-40cd-925d-ea512e11b1c4)

drgrice1 commented:
I am not seeing anything like what you show in your screenshot.  Have you run `npm ci` and restarted the server?  Are there any console log errors?

drgrice1 commented:
Here is what I see: 
![imageview-jsx](https://github.com/user-attachments/assets/0d331dc2-fd8b-4dfa-943a-a9b67b0f07d1)

somiaj commented:
Have run `npm ci` in `pg/htdocs` a few times, restarted the server (didn't think I needed to do that, but tried anyways on my dev server), and no errors in the console, though I do see 'MathJax (no yet) loaded', I don't think that is related.

I have also tested in both firefox-esr and chromium, same in both cases.

drgrice1 commented:
I have tested in firefox and crome.  I still don't see what you are seeing.  It doesn't happen on any problems.

drgrice1 commented:
By the way, the "MathJax (not yet) loaded" console log message generally always happens.  That is only about the initial graph on the page, and not the graph in the imageview dialog, and is not important.  It is just JSXGraph waiting for MathJax to load before typesetting and adjusting positioning of the typeset elements.

somiaj commented:
I was able to fix the issue by adding `jsxBoard?.zoom100();` right after the `jsxBoard?.resizeContainer` call. I see you messed with some stuff too.

drgrice1 commented:
The changes I made are unrelated to the issue you are seeing.  I want the code in the imageview.js file to be as unspecific to this as possible.  So I just moved the addition of the `jxgbox` and `plots-jsxgraph` classes to the JSXGraph.pm code instead of in the imageview.js file.

drgrice1 commented:
Do you have some other javascript running on the page or some other CSS loaded that might be interfering?

The `zoom100` call does not seem to be causing issues, but it just shouldn't be needed, and it isn't for me.  The graphtool uses the same technique when the page size changes also, and it certainly isn't needed there.  Something is clearly off with something you have.

somiaj commented:
I shouldn't have any custom css or javascript running on the problem I'm testing. I using the default `math4` theme, and have nothing in either of the `math4-overrides` files. Could there be some sort of race condition? I notice that sometimes when I load the page jsxgraph puts the label in the wrong location. Most of the times it is near the axis, but occasionally it is shifted up quite a bit.

![myscrot](https://github.com/user-attachments/assets/2525dfc9-9870-4d2d-ac45-f9ec6602eea4)

somiaj commented:
I added the `zoom100()` call since it resets the board to the original view port, which for some reason needs updated on my system, though I'm unsure what could be causing this, I just discovered that forcing a reset of the view port fixed the issue I was seeing, though it didn't identify why the view port wasn't be computed correctly in the first place.

drgrice1 commented:
That might make sense for the initial JSXGraph image in the page, but not for the image in the dialog.

Note that the `resizeContainer` call also resets the bounding box.  I examined the code of the two methods, and the `zoom100` call is just doing much the same as what the `resizeContainer` call does at the end.

drgrice1 committed:
Wait until the containing div has size before drawing the board.

This may fix the issue seen by @somiaj with the imageview dialog, but is
needed even if it doesn't because if the containing div does not have
size when the board is drawn the position of elements will be off, and
that will happen, for instance, if the div is inside a scaffold.

drgrice1 commented:
I added a change that might fix your issue.  Even if it doesn't the change is needed because, for instance, if the image is inside a scaffold that is initially closed, the containing div does not have size until the scaffold is opened.  As a result the elements will be out of position.

Test the last change with the following example.

```perl
DOCUMENT();
loadMacros('PGstandard.pl', 'PGML.pl', 'scaffold.pl', 'plots.pl');

$plot = Plot(
    xmin         => -1,
    xmax         =>  6,
    ymin         => -1,
    ymax         =>  7,
    xtick_delta  =>  1,
    ytick_delta  =>  1,
    xminor       =>  0,
    yminor       =>  0,
    xlabel       => '\(x \text{ (ft)}\)',
    ylabel       => '\(y \text{ (ft)}\)',
    xjsx_options => { label => { offset => [ -27, 5 ] } }
);
$plot->add_color('lightgreen', 156, 215, 151);
$plot->add_color('darkgreen',  0,   86,  34);
$plot->add_dataset(
    [ 0, 1 ], [ 3, 1 ], [ 3, 5 ], [ 0, 5 ], [ 0, 1 ],
    color      => 'darkgreen',
    fill       => 'self',
    fill_color => 'lightgreen'
);

for ('Tikz', 'JSXGraph') {
    $plot->image_type($_);
    $image{$_} = image($plot);
}

Scaffold::Begin();
Section::Begin('Tikz Image');
BEGIN_PGML
[$image{Tikz}]*
END_PGML
Section::End();

Section::Begin('JSXGraph Image');
BEGIN_PGML
[$image{JSXGraph}]*
END_PGML
Section::End();
Scaffold::End();

ENDDOCUMENT();
```

somiaj commented:
@drgrice1 that didn't fix my issue here.

drgrice1 commented:
It should fix the issue that you mentioned in https://github.com/somiaj/pg/pull/5#issuecomment-2963527332.

somiaj commented:
I am still seeing that issue too. I'm trying to update another dev server, to rule out it is a local issue.

somiaj commented:
Though I only get the issue when the image is outside of the scaffold. Once I put the image inside the scaffold as per your test problem it does go away.

drgrice1 committed:
Add a workaround for the issue with the bounding box when the zoom factor is not 100%.

Something somewhere in the code is messing with the bounding box, and I
am not sure what.  This does not happen with all of the other uses of
JSXGraph that I have.  Just this one.

drgrice1 commented:
I added a workaround for the issue for now.  I don't know why this is happening with the plot code.  This is the only usage of JSXGraph where this happens.  Something is causing the bounding box to be messed up.

somiaj commented:
That fixes the issue for me, anything else needed before I merge this?

drgrice1 commented:
For now it is good.  I am still not happy with the workaround, but it will do for now.
